fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/MetricUnionAvailability.java

```diff
@@ -0,0 +1,190 @@
+                                entry -> entry.getKey(),
+                                entry -> entry.getValue(),
+                                (value1, value2) -> {
+                                    List<Interval> combined = new ArrayList<>();
+                                    combined.addAll(value1);
+                                    combined.addAll(value2);
+                                    return combined;
+                                }
+                        )
+                );
+    }
+
+    @Override
+    public SimplifiedIntervalList getAvailableIntervals(DataSourceConstraint constraints) {
+        return new SimplifiedIntervalList(
+                constructSubConstraint(constraints).entrySet().stream()
+                        .map(entry -> entry.getKey().getAvailableIntervals(entry.getValue()))
+                        .map(i -> (Set<Interval>) new HashSet<>(i))
+                        .reduce(null, IntervalUtils::getOverlappingSubintervals)
+        );
+    }
+
+    /**
+     * Returns duplicate values of <tt>MetricColumn</tt>'s in a map of <tt>Availability</tt>
+     * to a set of <tt>MetricColumn</tt>'s contained in that <tt>Availability</tt>.
+     * <p>
+     * The return value is a map of duplicate <tt>MetricColumn</tt> to all <tt>Availability</tt>'s that contains
+     * this <tt>MetricColumn</tt>
+     *
+     * @param availabilitySetMap  A map from <tt>Availability</tt> to set of <tt>MetricColumn</tt> contained in that
+     * <tt>Availability</tt>
+     *
+     * @return duplicate values of <tt>MetricColumn</tt>'s
+     */
+    private static Map<MetricColumn, Set<Availability>> getDuplicateValue(
+            Map<Availability, Set<MetricColumn>> availabilitySetMap
+    ) {
+        Map<MetricColumn, Set<Availability>> metricColumnSetMap = new HashMap<>();
+
+        for (Map.Entry<Availability, Set<MetricColumn>> entry : availabilitySetMap.entrySet()) {
+            Availability availability = entry.getKey();
+            for (MetricColumn metricColumn : entry.getValue()) {
+                metricColumnSetMap.put(
+                        metricColumn,
+                        Sets.union(
+                                metricColumnSetMap.getOrDefault(metricColumn, new HashSet<>()),
+                                Sets.newHashSet(availability)
+                        )
+                );
+            }
+        }
+
+        return metricColumnSetMap.entrySet().stream()
+                .filter(entry -> entry.getValue().size() > 1)
+                .collect(
+                        Collectors.toMap(
+                                Map.Entry::getKey,
+                                Map.Entry::getValue
+                        )
+                );
+    }
+
+    /**
+     * Construct a mapping from a type of <tt>Availability</tt> to <tt>DataSourceConstraint</tt> of that availability
+     * and filter out mappings that maps to <tt>DataSourceConstraint</tt> with empty set of metric names.
```

Comment:
Which things are are getting filtered out? It's not clear what things might have empty metrics, nor under what conditions this might happen.